src/ui/public/vis/vis_type.js

```diff
@@ -15,7 +15,12 @@ export default function VisTypeFactory(Private) {
     this.schemas = opts.schemas || new VisTypeSchemas();
     this.params = opts.params || {};
     this.requiresSearch = opts.requiresSearch == null ? true : opts.requiresSearch; // Default to true unless otherwise specified
+    this.listeners = opts.listeners || {};
   }
 
+  VisType.prototype.createRenderbot = function (vis, $el, uiState) {
+    throw new Error('not implemented');
+  };
+
   return VisType;
 };
```

Comment:
I am not a 100% sure if we need to broker this with the renderbot at all, but maybe I miss something major.

Since every externally accessible visualization (instance of `vis`) needs to publish a 'renderComplete' event, I would propose modeling on the Vis-level.

E.g.:

1) We make `ui/vis/vis` support pub/sub (perhaps do the same for `ui/vis/vis_type` in fact)

```
 function Vis(indexPattern, state, uiState) {
    EventEmitter.call(this);
    ...  }
  Vis.prototype = Object.create(EventEmitter.prototype);
```

2) visualization implementations publish their readiness (templated vis, vislib vis)

e.g. in `metric_vis_controller`, (but also  `table_vis_controller`,  `vislib, ...)

```
      $scope.processTableGroups(tabifyAggResponse($scope.vis, resp));
      $scope.vis.emit('renderComplete');
    }
```

3) listeners can just subscribe to this events
e.g. in `ui/visualize/visualize`, (but also Reporting...)
```
 if (oldVis) $scope.renderbot = null;
 if (vis) {
      vis.on('renderComplete', () => {
                               //...here we could propage on the correct scope with $scope.$emit('renderComplete');
                              // ...but perhaps Reporting could  just listen to it directly
      });
  }
```

I think this will have several advantages:
- the implementation remains a little cleaner
   - we don't have clients conditionally implement a method. They can just use `on`.
   - publishers can just use emit() to publish, and don't have to worry if there is a registered listener or not
- Thinking longer ahead, this will be easier to document in JSDoc as well since @events are a core-type.